<?xml version="1.0"?>
<launch>

  <arg name="front_camera" default="tower_cam3d_front" />  
  <arg name="back_camera" default="tower_cam3d_back" />
  <arg name="barrier_tape_detection_config_file" default="$(find mir_barrier_tape_detection)/ros/config/barrier_tape_detection.yaml" />  
  
  <rosparam file="$(arg barrier_tape_detection_config_file)" command="load"/>
 
  <!-- Barrier tape detection node -->
  <group ns="mir_perception/front_camera">
      <node pkg="mir_barrier_tape_detection" type="barrier_tape_detection_node" name="barrier_tape_detection" output="screen" respawn="true">
          <remap from="~input_rgb_image" to="/$(arg front_camera)/rgb/image_raw"/>
          <remap from="~input_pointcloud" to="/$(arg front_camera)/depth_registered/points"/>
          <remap from="~camera_info" to="/$(arg front_camera)/rgb/camera_info"/>
          <param name="loop_rate" type="int" value="30" />
          <param name="target_frame" value="map"/>
          <remap from="~event_in" to="/mir_perception/barrier_tape_detection/event_in"/>
          <remap from="~output/yellow_barrier_tape_pointcloud" to="/mir_perception/barrier_tape_detection/output/yellow_barrier_tape_pointcloud"/>
      </node>
      <!-- Point cloud to laser scan conversion-->
      <node pkg="pointcloud_to_laserscan" type="pointcloud_to_laserscan_node" name="pointcloud_to_laserscan">
          <remap from="cloud_in" to="/mir_perception/barrier_tape_detection/output/yellow_barrier_tape_pointcloud"/>
          <remap from="scan" to="/barrier_tape/scan_front" />
          <rosparam>
            target_frame: base_laser_front_link # Leave disabled to output scan in pointcloud frame
            transform_tolerance: 0.01
            min_height: -1.0
            max_height: 5.0

            angle_min: -1.5708 # -M_PI/2
            angle_max: 1.5708 # M_PI/2
            angle_increment: 0.0087 # M_PI/360.0
            scan_time: 0.03333
            range_min: 0.001
            range_max: 5.0
            use_inf: false

            # Concurrency level, affects number of pointclouds queued for processing, thread number governed by nodelet manager
            # 0 : Detect number of cores
            # 1 : Single threaded
            # 2->inf : Parallelism level
            concurrency_level: 1
          </rosparam>
      </node>
  </group>

  <group ns="mir_perception/back_camera">
      
      <node pkg="mir_barrier_tape_detection" type="barrier_tape_detection_node" name="barrier_tape_detection"  output="screen" respawn="true">
          <remap from="~input_rgb_image" to="/$(arg back_camera)/rgb/image_raw"/>
          <remap from="~input_pointcloud" to="/$(arg back_camera)/depth_registered/points"/>
          <remap from="~camera_info" to="/$(arg back_camera)/rgb/camera_info"/>
          <param name="loop_rate" type="int" value="1" />
          <param name="target_frame" value="map"/>
          <remap from="~event_in" to="/mir_perception/barrier_tape_detection/event_in"/>
          <remap from="~output/yellow_barrier_tape_pointcloud" to="/mir_perception/barrier_tape_detection/output/yellow_barrier_tape_pointcloud"/>
      </node>

      <!-- Point cloud to laser scan conversion-->
      <node pkg="pointcloud_to_laserscan" type="pointcloud_to_laserscan_node" name="pointcloud_to_laserscan">
          <remap from="cloud_in" to="/mir_perception/barrier_tape_detection/output/yellow_barrier_tape_pointcloud"/>
          <remap from="scan" to="/barrier_tape/scan_rear" />
          <rosparam>
            target_frame: base_laser_rear_link # Leave disabled to output scan in pointcloud frame
            transform_tolerance: 0.01
            min_height: -1.0
            max_height: 5.0

            angle_min: -1.5708 # -M_PI/2
            angle_max: 1.5708 # M_PI/2
            angle_increment: 0.0087 # M_PI/360.0
            scan_time: 0.03333
            range_min: 0.001
            range_max: 5.0
            use_inf: false

            # Concurrency level, affects number of pointclouds queued for processing, thread number governed by nodelet manager
            # 0 : Detect number of cores
            # 1 : Single threaded
            # 2->inf : Parallelism level
            concurrency_level: 1
          </rosparam>
      </node>
  </group>
</launch>
